### ISSUE TITLE
[BUG][SPARK] ABFSS and WASBS report all files being used as Input Sources

### ISSUE THREAD

**Issue by wjohnson** [OpenLineage/OpenLineage#459]
When using OpenLineage's Spark Integration and using ABFSS or WASBS to connect to an Azure Data Lake Gen 2 / Blob Storage account, Open Lineage's Input data source will report back **all file paths being used by the spark job** rather than just the folder path in some scenarios.

The table below shows 12 scenarios testing ABFSS and WASBS vs Databricks Mount Points where we varied the file path input string to be either a single file, a folder path, or a folder path that uses a wildcard.  In addition we tested the read with inferSchema on or off which generates a second open lineage emit event.

| input source| single file| folder path | wildcard path | infer schema|
|--------------|-----------|--------------|----------------|---------------|
|wasbs|Returns the file path| Returns the folder path| Returns all file paths used|Returns all file paths|
| abfss| Returns the file path| Returns the folder path| Returns all file paths used|Returns all file paths|
| dbfs mount point| Returns the folder path| Returns the folder path| Returns the folder path|Returns the folder path|

We tested these scenarios using a scala spark command such as

```scala
val df = spark
.read
.format("csv")
.load(FILE_OR_FOLDER_PATH)
```

FILE_OR_FOLDER_PATH would take on a pattern such as:
* Single File: abfss://container@service.dfs.core.windows.net/folder/myfile.csv
* Folder Path: abfss://container@service.dfs.core.windows.net/folder/
* Wildcard Path: abfss://container@service.dfs.core.windows.net/folder/*

This is an example of what we mean by returning the file path and NOT the folder path.  It was generated by using `spark.read.format("csv").load("wasbs://CONTAINER@SERVICENAME.blob.core.windows.net/newretail/*")`

```json
{
    "inputs": [
        {
            "namespace": "wasbs://CONTAINER@SERVICENAME.blob.core.windows.net",
            "name": "/newretail/retail-1.csv",
            "facets": "..."
        },
        {
            "namespace": "wasbs://CONTAINER@SERVICENAME.blob.core.windows.net",
            "name": "/newretail/retail-2.csv",
            "facets": "..."
        }
    ]
}
```

**Desired / Expected Behavior**:

| input source| single file| folder path | wildcard path | infer schema|
|--------------|-----------|--------------|----------------|---------------|
|wasbs|:white_check_mark:Returns the file path| :white_check_mark:Returns the folder path| :collision:Returns the folder path| :collision:Returns the folder path|
| abfss|:white_check_mark:Returns the file path| :white_check_mark:Returns the folder path| :collision:Returns the folder path|:collision:Returns the folder path|
| dbfs mount point| :collision:Returns the file path| :white_check_mark:Returns the folder path| :white_check_mark:Returns the folder path|:white_check_mark:Returns the folder path|

* If a single file is mentioned, it would be ideal if Open Lineage always reported out that single file
* If wildcard path is mentioned, Open Lineage should only report back the appropriate parent folder
* InferSchema jobs should report only the single parent folder in all cases

**Comment by wjohnson** [comment]
I did some more digging on this one and it looks like there might be an issue due to `integration\spark\src\main\common\java\io\openlineage\spark\agent\util\PlanUtils.java` and the `getDirectoryPath` method doesn't know how to instantiate / use the ABFSS file path. It requires instantiating the filesystem with [ouath2 workflow](https://docs.microsoft.com/en-us/azure/databricks/data/data-sources/azure/adls-gen2/azure-datalake-gen2-sp-access#--dataframe-or-dataset-api) versus WASBS which follows an older convention and uses the `fs.azure.account.key` that PlanUtils seems to be looking for ([WASBS doc](https://docs.microsoft.com/en-us/azure/databricks/data/data-sources/azure/azure-storage#access-azure-blob-storage-directly)).

When comparing the logical plans between abfss and mount points, it's the exact same plan and the only difference is this warning it seems.

```
22/01/19 15:31:40 INFO OpenLineageRunEventBuilder: Visiting query plan Optional[== Parsed Logical Plan ==
InsertIntoHadoopFsRelationCommand abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput, false, CSV, [path=abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput/], Overwrite, [id, postalCode, streetAddress, city, stateAbbreviation]
+- Repartition 1, true
   +- Project [id#71, postalCode#72, streetAddress#73, city#78, stateAbbreviation#79]
      +- Join Inner, (id#71 = id#77)
         :- Relation[id#71,postalCode#72,streetAddress#73] csv
         +- Relation[id#77,city#78,stateAbbreviation#79] csv

== Analyzed Logical Plan ==

InsertIntoHadoopFsRelationCommand abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput, false, CSV, [path=abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput/], Overwrite, [id, postalCode, streetAddress, city, stateAbbreviation]
+- Repartition 1, true
   +- Project [id#71, postalCode#72, streetAddress#73, city#78, stateAbbreviation#79]
      +- Join Inner, (id#71 = id#77)
         :- Relation[id#71,postalCode#72,streetAddress#73] csv
         +- Relation[id#77,city#78,stateAbbreviation#79] csv

== Optimized Logical Plan ==
InsertIntoHadoopFsRelationCommand abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput, false, CSV, [path=abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput/], Overwrite, [id, postalCode, streetAddress, city, stateAbbreviation]
+- Repartition 1, true
   +- Project [id#71, postalCode#72, streetAddress#73, city#78, stateAbbreviation#79]
      +- Join Inner, (id#71 = id#77)
         :- Filter isnotnull(id#71)
         :  +- Relation[id#71,postalCode#72,streetAddress#73] csv
         +- Filter isnotnull(id#77)
            +- Relation[id#77,city#78,stateAbbreviation#79] csv

== Physical Plan ==
Execute InsertIntoHadoopFsRelationCommand abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput, false, CSV, [path=abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/exampleOutput/], Overwrite, [id, postalCode, streetAddress, city, stateAbbreviation]
+- AdaptiveSparkPlan isFinalPlan=true
   +- == Final Plan ==
      ShuffleQueryStage 1, Statistics(sizeInBytes=456.0 B, rowCount=5, isRuntime=true)
      +- Exchange RoundRobinPartitioning(1), REPARTITION_WITH_NUM, [id=#241]
         +- *(2) Project [id#71, postalCode#72, streetAddress#73, city#78, stateAbbreviation#79]
            +- *(2) BroadcastHashJoin [id#71], [id#77], Inner, BuildRight, false
               :- *(2) Filter isnotnull(id#71)
               :  +- FileScan csv [id#71,postalCode#72,streetAddress#73] Batched: false, DataFilters: [isnotnull(id#71)], Format: CSV, Location: InMemoryFileIndex[abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/example..., PartitionFilters: [], PushedFilters: [IsNotNull(id)], ReadSchema: struct<id:int,postalCode:string,streetAddress:string>
               +- BroadcastQueryStage 0, Statistics(sizeInBytes=1024.0 KiB, rowCount=5, isRuntime=true)
                  +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)),false), [id=#161]
                     +- *(1) Filter isnotnull(id#77)
                        +- FileScan csv [id#77,city#78,stateAbbreviation#79] Batched: false, DataFilters: [isnotnull(id#77)], Format: CSV, Location: InMemoryFileIndex[abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/example..., PartitionFilters: [], PushedFilters: [IsNotNull(id)], ReadSchema: struct<id:int,city:string,stateAbbreviation:string>
   +- == Initial Plan ==
      Exchange RoundRobinPartitioning(1), REPARTITION_WITH_NUM, [id=#78]
      +- Project [id#71, postalCode#72, streetAddress#73, city#78, stateAbbreviation#79]
         +- BroadcastHashJoin [id#71], [id#77], Inner, BuildRight, false
            :- Filter isnotnull(id#71)
            :  +- FileScan csv [id#71,postalCode#72,streetAddress#73] Batched: false, DataFilters: [isnotnull(id#71)], Format: CSV, Location: InMemoryFileIndex[abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/example..., PartitionFilters: [], PushedFilters: [IsNotNull(id)], ReadSchema: struct<id:int,postalCode:string,streetAddress:string>
            +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)),false), [id=#75]
               +- Filter isnotnull(id#77)
                  +- FileScan csv [id#77,city#78,stateAbbreviation#79] Batched: false, DataFilters: [isnotnull(id#77)], Format: CSV, Location: InMemoryFileIndex[abfss://rawdata@<MYADLSG2ACCT>.dfs.core.windows.net/examples/data/csv/example..., PartitionFilters: [], PushedFilters: [IsNotNull(id)], ReadSchema: struct<id:int,city:string,stateAbbreviation:string>
] with input dataset builders [io.openlineage.spark.agent.lifecycle.plan.LogicalRelationVisitor<org.apache.spark.sql.execution.datasources.LogicalRelation,D>, io.openlineage.spark.agent.lifecycle.plan.LogicalRDDVisitor<org.apache.spark.sql.execution.LogicalRDD,D>, io.openlineage.spark.agent.lifecycle.plan.BigQueryNodeVisitor<org.apache.spark.sql.catalyst.plans.logical.LogicalPlan,D>, io.openlineage.spark.agent.lifecycle.plan.KafkaRelationVisitor<org.apache.spark.sql.execution.datasources.LogicalRelation,D>, io.openlineage.spark.agent.lifecycle.plan.CommandPlanVisitor<org.apache.spark.sql.catalyst.plans.logical.LogicalPlan,io.openlineage.client.OpenLineage$InputDataset>]
22/01/19 15:31:40 WARN PlanUtils: Unable to get file system for path 
Failure to initialize configurationInvalid configuration value detected for fs.azure.account.key
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.services.SimpleKeyProvider.getStorageAccountKey(SimpleKeyProvider.java:51)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.AbfsConfiguration.getStorageAccountKey(AbfsConfiguration.java:552)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore.initializeClient(AzureBlobFileSystemStore.java:1679)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore.<init>(AzureBlobFileSystemStore.java:210)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.initialize(AzureBlobFileSystem.java:130)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2669)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)
	at io.openlineage.spark.agent.util.PlanUtils.getDirectoryPath(PlanUtils.java:159)
	at io.openlineage.spark.agent.lifecycle.plan.LogicalRelationVisitor.lambda$handleHadoopFsRelation$0(LogicalRelationVisitor.java:93)
        <SNIP>
	at io.openlineage.spark.agent.lifecycle.plan.LogicalRelationVisitor.handleHadoopFsRelation(LogicalRelationVisitor.java:101)
	at io.openlineage.spark.agent.lifecycle.plan.LogicalRelationVisitor.apply(LogicalRelationVisitor.java:71)
	at io.openlineage.spark.agent.lifecycle.plan.LogicalRelationVisitor.apply(LogicalRelationVisitor.java:48)
	at io.openlineage.spark.agent.util.PlanUtils$1.apply(PlanUtils.java:78)
        <SNIP>
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.lambda$visitLogicalPlan$13(OpenLineageRunEventBuilder.java:399)
	at io.openlineage.spark.agent.util.ScalaConversionUtils$3.apply(ScalaConversionUtils.java:126)
       <SNIP>
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.lambda$buildInputDatasets$5(OpenLineageRunEventBuilder.java:350)
	at java.util.Optional.map(Optional.java:215)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.buildInputDatasets(OpenLineageRunEventBuilder.java:348)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.populateRun(OpenLineageRunEventBuilder.java:307)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.buildRun(OpenLineageRunEventBuilder.java:293)
	at io.openlineage.spark.agent.lifecycle.OpenLineageRunEventBuilder.buildRun(OpenLineageRunEventBuilder.java:248)
	at io.openlineage.spark.agent.lifecycle.SparkSQLExecutionContext.end(SparkSQLExecutionContext.java:80)
	at io.openlineage.spark.agent.OpenLineageSparkListener.sparkSQLExecEnd(OpenLineageSparkListener.java:148)
        <SNIP>
Caused by: Invalid configuration value detected for fs.azure.account.key
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.diagnostics.ConfigurationBasicValidator.validate(ConfigurationBasicValidator.java:49)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.diagnostics.Base64StringConfigurationBasicValidator.validate(Base64StringConfigurationBasicValidator.java:40)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.services.SimpleKeyProvider.validateStorageAccountKey(SimpleKeyProvider.java:70)
	at shaded.databricks.azurebfs.org.apache.hadoop.fs.azurebfs.services.SimpleKeyProvider.getStorageAccountKey(SimpleKeyProvider.java:49)
	... 101 more
```

**Comment by mobuchowski** [comment]
@wjohnson is this comment for the second issue? https://github.com/OpenLineage/OpenLineage/issues/458

From what you posted before here, there is same behavior (file/directory/wildcard) between WASBS and ABFSS.

**Comment by collado-mike** [comment]
I see - we have this if condition to return the file's parent if we can detect it is a file, but that exception being thrown is preventing us from determining the type of the path - https://github.com/OpenLineage/OpenLineage/blob/main/integration/spark/src/main/common/java/io/openlineage/spark/agent/util/PlanUtils.java#L159 .

It looks like the `LogicalRelationVisitor` grabs the Hadoop `Configuration` object directly from the SparkContext. https://github.com/OpenLineage/OpenLineage/blob/main/integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/LogicalRelationVisitor.java#L91 . Based on the links you share, I would think the configuration should be in the configuration. I can try to replicate this (though I don't have an Azure Databricks account) and see if the configuration keys propagate

**Comment by collado-mike** [comment]
@wjohnson, can you try replacing the `handleHadoopFsRelation` method with the following code and try again?

This changes the `hadoopConfig` construction using the same code that the `InMemoryFileIndex` code uses, which means it should match the way those files are being read for processing. If it works for the execution, it should work for the metadata collection 🙏🏽
```java
  private List<D> handleHadoopFsRelation(LogicalRelation x) {
    HadoopFsRelation relation = (HadoopFsRelation) x.relation();
    return context.getSparkSession()
        .map(session -> {
          Configuration hadoopConfig = session.sessionState().newHadoopConfWithOptions(relation.options());
          return JavaConversions.asJavaCollection(relation.location().rootPaths()).stream()
              .map(p -> PlanUtils.getDirectoryPath(p, hadoopConfig))
              .distinct()
              .map(
                  p -> {
                    // TODO- refactor this to return a single partitioned dataset based on static
                    // static partitions in the relation
                    return datasetFactory.getDataset(p.toUri(), relation.schema());
                  })
              .collect(Collectors.toList());
        }).orElse(Collections.emptyList());
  }
```

**Comment by wjohnson** [comment]
@collado-mike BAM! That's it!  I no longer receive the PlanUtil exception and the behavior is now consistent across ABFSS, WASBS, and mount points in Azure Databricks.


| input source| single file| folder path | wildcard path | infer schema|
|--------------|-----------|--------------|----------------|---------------|
|wasbs|Returns the folder path| :white_check_mark:Returns the folder path| :white_check_mark:Returns the folder path| :white_check_mark:Returns the folder path|
| abfss|Returns the folder path| :white_check_mark:Returns the folder path| :white_check_mark:Returns the folder path|:white_check_mark:Returns the folder path|
| dbfs mount point|Returns the folder path| :white_check_mark:Returns the folder path| :white_check_mark:Returns the folder path|:white_check_mark:Returns the folder path|

**Comment by wjohnson** [comment]
Closing this issue since it was resolved in #491
